llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1777 @@
+
+
+
+			Value* ptr = load->getPointerOperand();
+			if (GetElementPtrInst* gep = dyn_cast<GetElementPtrInst>(ptr))
+			{
+				if (!gep->isInBounds())
+					return true;
+				//TODO: CHECK whether the location is also constant!!
+				return false;
+			}
+			//ConstantExpr are also possibly good
+			//PHI also
+			//
+			//In reality here we need a map of valid ptrs to be updated
+			//TODO: 
+			return false;
+		}	
+		return false;
+	}
+void clearFunction(llvm::Function& F)
+{
+	for (Instruction& I : instructions(F))
+	{
+		visitOuter<VisitingPolicy::REMOVE_VALUES>(I);
+	}
+}
+
+template <PartialInterpreter::VisitingPolicy policy>
+void visitOuter(llvm::Instruction& I) 
+	{
+			if (policy == REMOVE_VALUES)
+			{
+				if (computed.count(&I))
+					computed.erase(&I);
+				return;
+			}
+		
+		if (PHINode* phi = dyn_cast<PHINode>(&I))
+		{
+			const llvm::BasicBlock* from = getIncomingBB();
+			if (from)
+			{
+				llvm::Value* incoming = phi->getIncomingValueForBlock(from);
+				assert(incoming);
+				if (isValueComputed(incoming))
+					incomings.push_back({phi, {getOperandValue(incoming, getLastStack()), getBitMask(incoming)}});
+			}
+			return;
+		}
+		else
+		{
+			for (auto& p : incomings)
+			{
+		//		llvm::errs() <<"PHI-->\t"<< *p.first << "\t" ;
+		//		p.second.print("");
+				getLastStack().Values[const_cast<llvm::Value*>(p.first)] = p.second.first;
+				computed.insert(p.first);
+				stronglyKnownBits[const_cast<llvm::Value*>(p.first)] = p.second.second;
+			}
+			incomings.clear();
+		}
+///Here PHI have been properly processed
+
+		if (false)
+if (isa<CallBase>(&I) && sizeStack() == 1)
+{
+	CallBase* CB= dyn_cast<CallBase>(&I);
+	llvm::Function* pippo = CB->getCalledFunction();
+
+	if (pippo)
+	{
+
+		PartialInterpreter::DO_STUFF(pippo, CB, nullptr, *this);
+	}
+}
+
+
+		const bool skip = hasToBeSkipped(I);
+		const bool term = I.isTerminator();
+
+
+
+	if (false)
+{
+		if (skip || !term)
+		{
+			if (!skip)
+			llvm::errs() << "compute ";
+			else
+			llvm::errs() << "        ";
+		}
+		else
+			llvm::errs() << "COMPUTE ";
+		llvm::errs() << I << "\n";
+}
+
+		if (skip)
+		{
+
+			if (sizeStack() > 1)
+			{
+				popSingleStack();
+				//CALL FAILED, delete caller!
+				computed.erase(getLastStack().Caller);
+			}	
+			return; 
+		}
+
+		
+		if (term)
+		{
+
+		if (sizeStack() > 1)
+		{
+	//llvm::errs() << I << "\t" << I.getParent()->getParent()->getName() << "\tciccio\n";
+	BasicBlock* next = nullptr;
+	if (BranchInst* BI = dyn_cast<BranchInst>(&I))
+	{
+		if (BI->isConditional())
+		{
+			if (isValueComputed(BI->getCondition()))
+			{
+				GenericValue V = getOperandValue(BI->getCondition(), getLastStack());
+				if (V.IntVal == 0u)
+					next = BI->getSuccessor(1);
+				else
+					next = BI->getSuccessor(0);
+			}
+		}
+		else
+			next = BI->getSuccessor(0);
+	}
+	else if (ReturnInst* RI = dyn_cast<ReturnInst>(&I))
+	{
+		if (RI->getReturnValue())
+			stronglyKnownBits[getFirstStack().Caller] = getBitMask(RI->getReturnValue());
```

Comment:
How do we deal with multiple possible return sites? It would still be fine if all of them are the same value